osu.Desktop/Windows/WindowsAssociationManager.cs

```diff
@@ -0,0 +1,265 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
+using Microsoft.Win32;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Localisation;
+using osu.Framework.Logging;
+using osu.Game.Localisation;
+
+namespace osu.Desktop.Windows
+{
+    [SupportedOSPlatform("windows")]
+    public partial class WindowsAssociationManager : Component
+    {
+        public const string SOFTWARE_CLASSES = @"Software\Classes";
+
+        /// <summary>
+        /// Sub key for setting the icon.
+        /// https://learn.microsoft.com/en-us/windows/win32/com/defaulticon
+        /// </summary>
+        public const string DEFAULT_ICON = @"DefaultIcon";
+
+        /// <summary>
+        /// Sub key for setting the command line that the shell invokes.
+        /// https://learn.microsoft.com/en-us/windows/win32/com/shell
+        /// </summary>
+        public const string SHELL_OPEN_COMMAND = @"Shell\Open\Command";
+
+        public static readonly string EXE_PATH = Path.ChangeExtension(typeof(WindowsAssociationManager).Assembly.Location, ".exe");
+
+        /// <summary>
+        /// Program ID prefix used for file associations. Should be relatively short since the full program ID has a 39 character limit,
+        /// see https://learn.microsoft.com/en-us/windows/win32/com/-progid--key.
+        /// </summary>
+        public const string PROGRAM_ID_PREFIX = "osu";
+
+        private static readonly FileAssociation[] file_associations =
+        {
+            new FileAssociation(@".osz", WindowsAssociationManagerStrings.OsuBeatmap, Icons.Lazer),
+            new FileAssociation(@".olz", WindowsAssociationManagerStrings.OsuBeatmap, Icons.Lazer),
+            new FileAssociation(@".osr", WindowsAssociationManagerStrings.OsuReplay, Icons.Lazer),
+            new FileAssociation(@".osk", WindowsAssociationManagerStrings.OsuSkin, Icons.Lazer),
+        };
+
+        private static readonly UriAssociation[] uri_associations =
+        {
+            new UriAssociation(@"osu", WindowsAssociationManagerStrings.OsuProtocol, Icons.Lazer),
+            new UriAssociation(@"osump", WindowsAssociationManagerStrings.OsuMultiplayer, Icons.Lazer),
+        };
+
+        [Resolved]
+        private LocalisationManager localisation { get; set; } = null!;
+
+        private IBindable<LocalisationParameters> localisationParameters = null!;
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            localisationParameters = localisation.CurrentParameters.GetBoundCopy();
+            InstallAssociations();
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            localisationParameters.ValueChanged += _ => updateDescriptions();
```

Comment:
It is a bit veiled, but what this ends up implying is that every single change of user language, or even the "show metadata in original language" toggle in settings, _is going to cause writes to the windows registry_. I am not comfortable with this and do not think that is sane.

I think there are two paths I'd rather take here:
- not bother with localising descriptions at all
- localise them, but on a one-shot basis, at point of installation.

Reply:
> This relies too heavily on the automatic language selection being correct

What do you mean by "too heavily"? I'm not sure I've known that to fail that much. If there's one place where that should work perfectly it's windows.

> How about updating the localised description once the user selects and confirms their language in the first-run setup wizard?

Eeeeehhhh I don't know. I'd rather not touch the registry outside of squirrel flows.

Reply:
@ppy/team-client I'm happy with the code and this PR is ready for review. The above conversation remains the only open question.